python/lsst/daf/butler/core/databaseDict.py

```diff
@@ -0,0 +1,101 @@
+# This file is part of daf_butler.
+#
+# Developed for the LSST Data Management System.
+# This product includes software developed by the LSST Project
+# (http://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+from collections.abc import MutableMapping
+
+from .utils import doImport
+
+
+class DatabaseDict(MutableMapping):
+    """An abstract base class for dict-like objects backed by a database.
+
+    DatabaseDict subclasses must implement the abstract ``__getitem__``,
+    ``__setitem__``, ``__delitem__`, ``__iter__``, and ``__len__`` abstract
+    methods defined by `MutableMapping`.
+
+    They must also provide a constructor that takes the same arguments as that
+    of `DatabaseDict` itself, *unless* they are constructed solely by
+    `Registry.makeDatabaseDict` (in which case any constructor arguments are
+    permitted).
+
+    Parameters
+    ----------
+    config : `Config`
+        Configuration used to identify and construct a subclass.
+    fields : `dict`
+        A dictionary mapping `str` field names to type objects, containing
+        all fields to be held in the database.
+    key : `str` or `tuple`
+        The name of the field to be used as keys in the dictionary, or
+        a tuple thereof.
+    value : `type`
+        The type used for the dictionary's values, typically a `namedtuple`.
+        Must have a ``_field`` class attribute that is a tuple of field
```

Comment:
Is this `_fields` rather than `_field`? I found this all a bit confusing until I worked out that `_fields` and `_make` came from `namedtuple`.